Prompt user to re-register if the authorization fails

whot opened the pull request:
A bit messy because I'm sure half-way through some of the bits get broken because our messages don't all have errnos anymore. But... meh. It'll work in the final product.

The main thing this does after shuffling the exception handling around is to hook up the GUI with the `AuthorizationError`. It shows an overlay with a button to click to re-invoke the setup process.

Easiest way to test: `./builddir/tuhi.devel --config-dir="/tmp/$(date -Idate)" --peek`, register the device, then run the normal `./builddir/tuhi.devel` again.

whot committed:
protocol: implement the message to wait for the end of pen data

This pushes the CRC fetching into the messages, making the caller code more
generic.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: use the DeviceError for the invalid state message

Nothing throws the previous exception anymore anyway.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: use a MisingReplyError instead of timeout exceptions

It's caused by a timeout but the underlying issue is that we never got the
reply that we wanted.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom.py: drop two now-unused functions

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: replace one 'old' exception with the new protocol one

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: replace the last exception with a protocol exception

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
dbus: change EAGAIN to EBUSY for "already listening"

Seems like a more precise description.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: augment the new exceptions with errnos

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
gui: pass a dbus device error on as signal

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: move an assignment to a shared path

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: print the DeviceError as something useful

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: on the spark treat a general error during connect as wrong uuid

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
gui: hook up the auth signal with a re-register workflow

When we get an authorization error, show a revealer to re-register the device.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

bentiss commented:
Hmm, doesn't seem to be working here:
- `/builddir/tuhi.devel --config-dir="/tmp/$(date -Idate)" --peek`
  -> all fine, it re-registers
- `/builddir/tuhi.devel --peek`
  -> exception:
```shell
09:42:02 DEBUG: tuhi: E0:61:C6:BF:14:4E: UUID 3503516e7e74 protocol: slate
09:42:02 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connecting
09:42:02 DEBUG: tuhi: E0:61:C6:BF:14:4E: UUID 3503516e7e74 protocol: slate
09:42:02 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connecting
09:42:02 DEBUG: tuhi.ble.E0:61:C6:BF:14:4E: Already connecting
09:42:02 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connection established
09:42:04 DEBUG: tuhi: E0:61:C6:BF:14:4E: connected for DeviceMode.LISTEN
09:42:04 DEBUG: tuhi.wacom: glamorous sheep is using protocol SLATE
09:42:04 DEBUG: tuhi.wacom: E0:61:C6:BF:14:4E: starting
09:42:04 DEBUG: tuhi.fw: command: check connection
09:42:04 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> e6 / 06 / 35 03 51 6e 7e 74
09:42:04 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 07
09:42:04 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Disconnecting
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 1015, in _run
    self._wacom_protocol.retrieve_data()
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 808, in retrieve_data
    self.check_connection()
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 509, in check_connection
    self.p.execute(Interactions.CONNECT, self._uuid)
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 285, in execute
    return self.get(key, *args, **kwargs).execute()
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 554, in execute
    self._handle_reply(self.reply)
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 607, in _handle_reply
    super()._handle_reply(reply)
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 535, in _handle_reply
    raise DeviceError(reply[0])
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 442, in __init__
    self.errorcode = DeviceError.ErrorCode(errorcode)
  File "/usr/lib64/python3.7/enum.py", line 309, in __call__
    return cls.__new__(cls, value)
  File "/usr/lib64/python3.7/enum.py", line 543, in __new__
    return cls._missing_(value)
  File "/usr/lib64/python3.7/enum.py", line 556, in _missing_
    raise ValueError("%r is not a valid %s" % (value, cls.__name__))
ValueError: 7 is not a valid ErrorCode

09:42:04 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Disconnected
09:42:04 DEBUG: tuhi: E0:61:C6:BF:14:4E: disconnected
```

The spec says error code 7  is deprecated, but there might be firmawres around that still use those, so we should not bail out on deprecated error codes.

whot commented:
fully agree. fwiw I started with implementing the spec with the intention of adding the older parts later but given the "accuracy" of the spec I forgot about those while battling the devices that's supposed to work correctly...

whot committed:
protocol: handle error code 0x7

Deprecated, but still exists on the Slate so let's wrap the connect attempt
so we throw the right error.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot commented:
Fixed with the last commit treating `0x7` as `AuthorizationError` during connect

bentiss reviewed (approved):
LGTM now
